langchain4j-weaviate/src/main/java/dev/langchain4j/store/embedding/WeaviateEmbeddingStoreImpl.java

```diff
@@ -0,0 +1,192 @@
+  private static final Double DEFAULT_MIN_CERTAINTY = 0.8;
+  private static final String METADATA_TEXT_SEGMENT = "text";
+  private static final String ADDITIONALS = "_additional";
+
+  private final WeaviateClient client;
+  private final String objectClass;
+
+  @Builder
+  @SneakyThrows
+  public WeaviateEmbeddingStoreImpl(String apiKey, String scheme, String host, String objectClass) {
+    client = WeaviateAuthClient.apiKey(new Config(scheme, host), apiKey);
+    this.objectClass = objectClass != null ? objectClass : DEFAULT_CLASS;
+  }
+
+  @Override
+  public String add(Embedding embedding) {
+    val id = generateRandomId();
+    add(id, embedding);
+    return id;
+  }
+
+  @Override
+  public void add(String id, Embedding embedding) {
+    addAll(singletonList(id), singletonList(embedding), null);
+  }
+
+  @Override
+  public String add(Embedding embedding, TextSegment textSegment) {
+    return addAll(singletonList(embedding), textSegment == null ? null : singletonList(textSegment))
+      .stream()
+      .findFirst()
+      .orElse(null);
+  }
+
+  @Override
+  public List<String> addAll(List<Embedding> embeddings) {
+    return addAll(embeddings, null);
+  }
+
+  @Override
+  public List<String> addAll(List<Embedding> embeddings, List<TextSegment> embedded) {
+    return addAll(null, embeddings, embedded);
+  }
+
+  @Override
+  public List<EmbeddingMatch<TextSegment>> findRelevant(Embedding referenceEmbedding, int maxResults) {
+    return findRelevant(referenceEmbedding, maxResults, DEFAULT_MIN_CERTAINTY);
+  }
+
+  @Override
+  public List<EmbeddingMatch<TextSegment>> findRelevant(
+    Embedding referenceEmbedding,
+    int maxResults,
+    double minCertainty
+  ) {
+    val result = client
+      .graphQL()
+      .get()
+      .withClassName(objectClass)
+      .withFields(
+        Field.builder().name(METADATA_TEXT_SEGMENT).build(),
+        Field
+          .builder()
+          .name(ADDITIONALS)
+          .fields(
+            Field.builder().name("id").build(),
+            Field.builder().name("certainty").build(),
+            Field.builder().name("vector").build()
+          )
+          .build()
+      )
+      .withNearVector(
+        NearVectorArgument
+          .builder()
+          .vector(referenceEmbedding.vectorAsList().toArray(new Float[0]))
+          .certainty((float) minCertainty)
+          .build()
+      )
+      .withLimit(maxResults)
+      .run();
+
+    if (result.hasErrors()) {
+      return emptyList();
+    }
+
+    val resGetPart = ((Map<String, Map>) result.getResult().getData()).entrySet().stream().findFirst();
+    if (!resGetPart.isPresent()) {
+      return emptyList();
+    }
+
+    val resItemsPart = resGetPart.get().getValue().entrySet().stream().findFirst();
+    if (!resItemsPart.isPresent()) {
+      return emptyList();
+    }
+
+    val resItems = ((Map.Entry<String, List<Map<String, ?>>>) resItemsPart.get()).getValue();
```

Comment:
I can imagine your pain when working with this API😂

Reply:
yes, it was really challenging 🤣